include/tapkee/methods.hpp

```diff
@@ -22,58 +22,53 @@ class DynamicImplementation : public ImplementationBase<RandomAccessIterator, Ke
 {
 public:
     using ImplementationBase<RandomAccessIterator, KernelCallback, DistanceCallback, FeaturesCallback>::ImplementationBase;
-    TapkeeOutput embedUsing(DimensionReductionMethod method)
+    TapkeeOutput embedUsing(const DimensionReductionMethod& method)
     {
+        timed_context tctx__(fmt::format("[+] embedding with {}", method.name()));
+
         if (this->context.is_cancelled())
             throw cancelled_exception();
 
+        if (method.needs_kernel && is_dummy<KernelCallback>::value)
+        {
+            throw unsupported_method_error("Kernel callback is missed");
+        }
+        if (method.needs_distance && is_dummy<DistanceCallback>::value)
+        {
+            throw unsupported_method_error("Distance callback is missed");
+        }
+        if (method.needs_features && is_dummy<FeaturesCallback>::value)
+        {
+            throw unsupported_method_error("Features callback is missed");
+        }
+
 #define tapkee_method_handle(X)                                                                                        \
-    case X: {                                                                                                          \
-        timed_context tctx__("[+] embedding with " #X);                                                                \
-        if (MethodTraits<X>::needs_kernel && is_dummy<KernelCallback>::value)                                          \
-        {                                                                                                              \
-            throw unsupported_method_error("Kernel callback is missed");                                               \
-        }                                                                                                              \
-        if (MethodTraits<X>::needs_distance && is_dummy<DistanceCallback>::value)                                      \
-        {                                                                                                              \
-            throw unsupported_method_error("Distance callback is missed");                                             \
-        }                                                                                                              \
-        if (MethodTraits<X>::needs_features && is_dummy<FeaturesCallback>::value)                                      \
-        {                                                                                                              \
-            throw unsupported_method_error("Features callback is missed");                                             \
-        }                                                                                                              \
+    if (method.is(X)) {                                                                                                \
```

Comment:
Could this be `method == X` so that `Method::is(const Method&)` is not needed? Just wild-asking, it can be that the method is required in other parts of the code. I was thinking toward removing `Method` or at least the inheritance relationship there for a bit more of code reduction.